tasks/broad/RNAWithUMIsTasks.wdl

```diff
@@ -1,617 +1,628 @@
+  }
 }
 
 task SortSamByQueryName {
-    input {
-        File input_bam
-        String output_bam_basename
-
-        # SortSam spills to disk a lot more because we are only store 300000 records in RAM now because its faster for our data so it needs
-        # more disk space.  Also it spills to disk in an uncompressed format so we need to account for that with a larger multiplier
-        Float sort_sam_disk_multiplier = 6.0
-
-        String docker = "us.gcr.io/broad-gotc-prod/picard-cloud:2.26.6"
-        Int cpu = 1
-        Int memory_mb = 7500
-        Int disk_size_gb = ceil(sort_sam_disk_multiplier * size(input_bam, "GiB")) + 20
-    }
+  input {
+    File input_bam
+    String output_bam_basename
 
-    command <<<
-        java -Xms6500m -Xmx7000m  -jar /usr/picard/picard.jar \
-        SortSam \
-        INPUT=~{input_bam} \
-        OUTPUT=~{output_bam_basename}.bam \
-        SORT_ORDER="queryname" \
-        CREATE_MD5_FILE=true \
-        MAX_RECORDS_IN_RAM=300000
-    >>>
-
-    runtime {
-        docker: docker
-        cpu: cpu
-        memory: "${memory_mb} MiB"
-        disks : "local-disk ${disk_size_gb} HDD"
-    }
+    # SortSam spills to disk a lot more because we are only store 300000 records in RAM now because its faster for our data so it needs
+    # more disk space.  Also it spills to disk in an uncompressed format so we need to account for that with a larger multiplier
+    Float sort_sam_disk_multiplier = 6.0
 
-    output {
-        File output_bam = "~{output_bam_basename}.bam"
-        File output_bam_md5 = "~{output_bam_basename}.bam.md5"
-    }
+    String docker = "us.gcr.io/broad-gotc-prod/picard-cloud:2.26.6"
+    Int cpu = 1
+    Int memory_mb = 7500
+    Int disk_size_gb = ceil(sort_sam_disk_multiplier * size(input_bam, "GiB")) + 20
+  }
+
+  Int java_memory_size = memory_mb - 1000
+  Int max_heap = memory_mb - 500
+
+  command <<<
+    java -Xms~{java_memory_size}m -Xmx~{max_heap}m -jar /usr/picard/picard.jar SortSam \
+      INPUT=~{input_bam} \
+      OUTPUT=~{output_bam_basename}.bam \
+      SORT_ORDER="queryname" \
+      CREATE_MD5_FILE=true \
+      MAX_RECORDS_IN_RAM=300000
+  >>>
+
+  runtime {
+    docker: docker
+    cpu: cpu
+    memory: "${memory_mb} MiB"
+    disks: "local-disk ${disk_size_gb} HDD"
+  }
+
+  output {
+    File output_bam = "~{output_bam_basename}.bam"
+    File output_bam_md5 = "~{output_bam_basename}.bam.md5"
+  }
 }
 
 task GroupByUMIs {
-    input {
-        File bam
-        File bam_index
-        String output_bam_basename
-
-        String docker = "us.gcr.io/broad-gotc-prod/umi_tools:1.0.0-1.1.1-1638821470"
-        Int cpu = 8
-        Int memory_mb = 64000
-        Int disk_size_gb = ceil(2.2 * size(bam, "GB")) + 300
-    }
+  input {
+    File bam
+    File bam_index
+    String output_bam_basename
+
+    String docker = "us.gcr.io/broad-gotc-prod/umi_tools:1.0.0-1.1.1-1638821470"
+    Int cpu = 2
+    Int memory_mb = 7500
```

Comment:
assuming the original memory allocation was a huge overkill?

Reply:
Yeah, I looked at these tasks (after running the scientific test) using the output of the memory profiling tool and found I could reduce it significantly

Reply:
Was this the monitoring script or some other tool? (just curious)

Reply:
Yes, after running the scientific tests, I checked the monitoring log for memory usage.